src/db/util/conditions.ts

```diff
@@ -0,0 +1,303 @@
+/* eslint-disable @typescript-eslint/no-namespace */
+// we use namespaces here to neatly organize and scope the many type and type guard definitions
+
+import Knex = require('knex');
+
+/**
+ * Symbols to use for where condition construction
+ */
+export namespace ConditionSymbols {
+  export const BUILDER = Symbol('builder');
+  export const AND = Symbol('and');
+  export const OR = Symbol('or');
+
+  /**
+   * Symbols to use for where condition construction
+   */
+  export const Cond = {
+    BUILDER: BUILDER,
+    AND: AND,
+    OR: OR,
+  } as const;
+}
+
+export const Cond = ConditionSymbols.Cond;
+
+/**
+ * Symbols to use when constructing conditions for a single property
+ */
+export namespace PropertySymbols {
+  export const IN = Symbol('in');
+  export const NOT_IN = Symbol('not in');
+  export const IS_NULL = Symbol('is null');
+  export const BETWEEN = Symbol('between');
+  export const LIKE = Symbol('like');
+  export const ILIKE = Symbol('ilike');
+  export const LT = Symbol('less than');
+  export const LTE = Symbol('less than or equal to');
+  export const GT = Symbol('greater than');
+  export const GTE = Symbol('greater than or equal to');
+
+  /**
+   * Symbols to use when constructing conditions for a single property
+   */
+  export const Op = {
+    IN: IN,
+    NOT_IN: NOT_IN,
+    IS_NULL: IS_NULL,
+    BETWEEN: BETWEEN,
+    LIKE: LIKE,
+    ILIKE: ILIKE,
+    LT: LT,
+    LTE: LTE,
+    GT: GT,
+    GTE: GTE,
+  } as const;
+}
+
+export const Op = PropertySymbols.Op;
+
+namespace PropertyConditions {
+  export type EqualityCondition<T> = T;
+
+  export type InCondition<T> = {
+    [Op.IN]: Iterable<T>;
+  };
+  export type NotInCondition<T> = {
+    [Op.NOT_IN]: Iterable<T>;
+  };
+
+  export type IsNullCondition = {
+    [Op.IS_NULL]: boolean;
```

Comment:
I was thinking that we should use `{propertyName: null}` for `propertyName IS NULL` and introduce `[Op.IS_NOT_NULL]: null` or `[Op.NE]: null`, but I am fine with what you used.

It is saying that `NULL` should be treated as a special case, and even though there are two ways to write `IS NULL` now (`{propertyName: null}` is more likely to be used), let's not complicate any further and use what you suggested.